mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +185,439 @@ def _chain_with_retriever(model):
+            run_id: Id of the run to log the artifacts.
+        """
+
+        def __init__(self):
+            super().__init__()
+            self._mlflow_client = MlflowClient()
+            self._run_span_mapping: Dict[str, MLflowSpanWrapper] = {}
+
+        def _get_span_by_run_id(self, run_id: UUID) -> Optional[MLflowSpanWrapper]:
+            if span := self._run_span_mapping.get(str(run_id)):
+                return span
+            raise MlflowException(f"Span for run_id {run_id!s} not found.")
+
+        def _start_span(
+            self,
+            span_name: str,
+            parent_run_id: Optional[UUID],
+            span_type: str,
+            run_id: UUID,
+            inputs: Optional[Dict[str, Any]] = None,
+            attributes: Optional[Dict[str, Any]] = None,
+        ) -> MLflowSpanWrapper:
+            """Start MLflow Span (or Trace if it is root component)"""
+            parent = self._get_span_by_run_id(parent_run_id) if parent_run_id else None
+            if parent:
+                span = self._mlflow_client.start_span(
+                    name=span_name,
+                    request_id=parent.request_id,
+                    parent_span_id=parent.span_id,
+                    span_type=span_type,
+                    inputs=inputs,
+                    attributes=attributes,
+                )
+            else:
+                # When parent_run_id is None, this is root component so start trace
+                span = self._mlflow_client.start_trace(
+                    name=span_name, inputs=inputs, attributes=attributes
+                )
+            self._run_span_mapping[str(run_id)] = span
+            return span
+
+        def _end_span(
+            self,
+            span: MLflowSpanWrapper,
+            outputs=None,
+            attributes=None,
+            status=SpanStatus(TraceStatus.OK),
+        ):
+            """Close MLflow Span (or Trace if it is root component)"""
+            self._mlflow_client.end_span(
+                request_id=span.request_id,
+                span_id=span.span_id,
+                outputs=outputs,
+                attributes=attributes,
+                status=status,
+            )
+
+        def _reset(self):
+            self._run_span_mapping = {}
+
+        @override
+        def on_chat_model_start(
+            self,
+            serialized: Dict[str, Any],
+            messages: List[List[BaseMessage]],
+            *,
+            run_id: UUID,
+            tags: Optional[List[str]] = None,
+            parent_run_id: Optional[UUID] = None,
+            metadata: Optional[Dict[str, Any]] = None,
+            name: Optional[str] = None,
+            **kwargs: Any,
+        ):
+            """Run when a chat model starts running."""
+            if metadata:
+                kwargs.update({"metadata": metadata})
+            llm_inputs = {"messages": [[dumpd(msg) for msg in batch] for batch in messages]}
```

Comment:
Now our JSON serializer supports pydantic model (perhaps can extend specifically for LC models tho), can we leave this field unserialized until logging? Schema-wise `Span` should have non-serialized inputs and outputs.

Reply:
Seems like this leads to error when rendering 
<img width="1065" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/d46a7336-2d10-462f-8425-e1912cde7866">
<img width="909" alt="image" src="https://github.com/mlflow/mlflow/assets/82044803/335374da-0513-4e39-84cd-d549ebf4c6f3">

Let me keep this for now and after UI fixes we can update. The `dumpd` here is actually returning a dictionary with json-serializable fields, so we should be safe here.

Reply:
@serena-ruan It seems the error comes from `attributes` rather then `inputs`. I tried logging without setting attributes in `on_chat_model_start` and it renders correctly (also without input serialization). 

The issue looks like handling of nested attributes. We can check this by the simple example below:
```
with mlflow.start_span(name="span", attributes={"foo": {"bar": "baz"}}):
    pass
```

FYI:
* ```attributes={"foo": None}``` -> This works fine.
* ```attributes={"foo": "bar", "baz": 1}``` -> This works fine.
* ```attributes={'foo': {'bar': 'baz'}}``` -> This fails.

cc: @daniellok-db